File: official/trace/optrace/sample/sample.cc

#include "optrace.h"
#include <iostream>
#include <memory>
#include <cstring>

using namespace OPTRACE;

//executing sample:
// 1. op-trace "A123:0:0 %3:<32x16xf16> = torch.2_10.aten::add(%1:<32x16xf16>{1, 32}+1024, %2:<f16>, 2.0:f32)"
// 2. op-trace optrace_sample.log
int main(int argc, char **argv)
{
  if (argc <= 1) {
    std::cout << "please provide log string or file" << std::endl;
    return 0;
  }

  size_t datalen = strlen(argv[1]);
  if (datalen <= 4) {
    std::cout << "bad input, data length less than 4 bytes, can't check .log postfix." << std::endl;
    return 0;
  }

  std::unique_ptr<optrace> optracer;

  if (argv[1][datalen - 1] == 'g' && argv[1][datalen - 2] == 'o'
    && argv[1][datalen - 3] == 'l' && argv[1][datalen - 4] == '.') {
    // filename
    optracer = std::make_unique<optrace>(argv[1]);
  }
  else {
    // log string
    optracer = std::make_unique<optrace>(argv[1], static_cast<int>(strlen(argv[1])));
  }

  auto m1 = optracer->getModule();
  if (m1.get() == nullptr) {
    std::cout << "failure get module." << std::endl;
    return -1;
  }

  // copy module test
  OPTRACE::module* m = new OPTRACE::module(*m1.get());

  // dump instruct by getString() interface.
  std::cout << "-------------------------------------------" << std::endl;
  std::cout << "getString by objects: " << std::endl;
  std::cout << m->getString();
  std::cout << std::endl;
  std::cout << "-------------------------------------------" << std::endl << std::endl;

  // access instruct object interfaces.
  int instructIndex = 0;
  // iterator all instructs
  for (auto& instructIter : m->getInstructs()) {
    std::cout << "instruct: " << instructIndex ++ << std::endl;
    std::cout << "\tprocessid: " << instructIter->getProcessID() << ", "
      << "rank:" << instructIter->getRankID() << ", "
      << "stream:" << instructIter->getStreamID() << std::endl;

    // iterator all output results
    std::cout << "\tresults:  \n";
    for (auto& resultIter : instructIter->getResults()) {
      auto type = resultIter->getValueType();
      // dynamic cast operand object to tensor object
      if (resultIter->getValueType() == valuetype::TENSOR) {
        tensor* t = dynamic_cast<tensor*>(resultIter.get());
        std::cout << "\t\ttensor: "
          << "tensorid:" << t->getTensorID() << ", "
          << "rank:" << t->getRank() << ", "
          << "type:" << static_cast<int>(t->getDataType()) << ", "
          << "dims:[";
        for (auto dimiter : t->getDims()) {
          std::cout << dimiter << ",";
        }
        std::cout << "], "
          << "strides:[";
        for (auto strideIter : t->getStrides()) {
          std::cout << strideIter << ",";
        }
        std::cout << "]" << std::endl;
      }
      // dynamic cast operand object to scalar object
      else if (resultIter->getValueType() == valuetype::SCALAR) {
        scalar* t = dynamic_cast<scalar*>(resultIter.get());
        std::cout << "\t\tscalar: "
          << "tensorid:" << t->getTensorID() << ", "
          << "name:" << t->getName() << ", "
          << "type:" << static_cast<int>(t->getDataType()) << ", "
          << "value:" << t->getValue() << std::endl;
      }
      // dynamic cast operand object to structure object
      else if (resultIter->getValueType() == valuetype::STRUCT) {
        structure* t = dynamic_cast<structure*>(resultIter.get());
        std::cout << "\t\tstructure: "
          << "tensorid name:" << t->getTensorIDName() << ", "
          << "tensorid: " << t->getTensorID() << ", "
          << "name: " << t->getName() << std::endl;
      }
    }

    // iterator all input operands
    for (auto& operandIter : instructIter->getOperands()) {
      // dynamic cast operand object to tensor object
      if (operandIter->getValueType() == valuetype::TENSOR) {
        tensor* t = dynamic_cast<tensor*>(operandIter.get());
        // same with result
      }
      // dynamic cast operand object to scalar object
      else if (operandIter->getValueType() == valuetype::SCALAR) {
        scalar *t = dynamic_cast<scalar*>(operandIter.get());
        // same with result
      }
      // dynamic cast operand object to structure object
      else if (operandIter->getValueType() == valuetype::STRUCT) {
        structure* t = dynamic_cast<structure*>(operandIter.get());
        // same with result
      }
    }
  }
}

File: official/trace/optrace/sample/sample.py
import sys
import torchtrace.optrace.optrace_py as optrace
import os

print("Current working directory:", os.getcwd())

expected_count = 2  # 期望的输入个数
input = ""
if len(sys.argv) == expected_count:
    input = sys.argv[1]
else:
    raise ValueError(f"Expected {expected_count} inputs, but got {len(sys.argv)}.")

isfile = False
if ".log" in input:
    isfile = True

opt = None
if isfile:
    opt = optrace.optrace(input)
else:
    opt = optrace.optrace(input, len(input))

mod = opt.getModule()

# string to object sample
print("\n\nstring to object sample")
instruct_list = mod.getInstructs()

for instr_dump in instruct_list:
    print(instr_dump.getString())

for instr in instruct_list:
    print("INSTRUCT processid: %s, rank: %d, stream: %d, op: %s" %(instr.getProcessID(), instr.getRankID(), instr.getStreamID(), instr.getOpname()))

    for val in instr.getResults():
        if isinstance(val, optrace.tensor):
            dimstr = ""
            for d in val.getDims():
                dimstr += str(d)
                dimstr += ", "
            print("    tensor: tensor id: %d, rank: %d, type:%d, dims:[%s]" % (val.getTensorID(), val.getRank(), val.getDataType(), dimstr))
        elif isinstance(val, optrace.scalar):
            print("    scalar: tensor id: %d, name: %s, type:%d, value:%s" % (val.getTensorID(), val.getName(), val.getDataType(), val.getValue()))
        elif isinstance(val, optrace.structure):
            print("    structure: tensor name: %s, tensor id: %d, name:%s" % (val.getTensorIDName(), val.getTensorID(), val.getName()))
        else:
            print("dynamic cast failure")


# object to string sample
print("\n\nobject to string sample")

# create customer type scalar:  name + data type + value + type string
scalar_0 = optrace.scalar("customertype", optrace.datatype.CUSTOM_DATA_TYPE, "none", "NoneType")
print(scalar_0.getString())

# create scalar: tensorID + data type
scalar_1 = optrace.scalar(1, optrace.datatype.I32)
print(scalar_1.getString())

# crate scalar: name + data type + value
scalar_2 = optrace.scalar(-1, "offset", optrace.datatype.I32, "1024")
print(scalar_2.getString())
# crate scalar: name + data type + value
scalar_3 = optrace.scalar("offset", optrace.datatype.I32, "1024")
print(scalar_3.getString())

#create tensor: tensorID + data type + dims + strides + offset
tensor_1 = optrace.tensor(2, optrace.datatype.I32, [2, 3, 4], [12, 4, 1], 512)
print(tensor_1.getString())
#create tensor: tensorID + data type + dims
tensor_2 = optrace.tensor(3, optrace.datatype.I32, [2, 3, 4])
print(tensor_2.getString())
#create tensor: tensorID + data type + dims + offset
tensor_3 = optrace.tensor(4, optrace.datatype.I32, [2, 3, 4], [], 512)
print(tensor_3.getString())

#create instruct
instruct_1 = optrace.instruct("A123", 0, 0, "aten::add", "torch", "2_11", [tensor_1, scalar_1, scalar_2], [tensor_2])
print(instruct_1.getString())

#create structure tensor + scalare
struct_1 = optrace.structure(5, "tuple", [scalar_2, tensor_1, scalar_1])
print(struct_1.getString())
#create structure tensor + scalar + structure
struct_2 = optrace.structure(6, "tuple", [struct_1, tensor_1, scalar_1])
print(struct_2.getString())
#create structure tensor + scalar + parameter named structure
struct_3 = optrace.structure("tuple07", 7, "tuple", [struct_1, tensor_1, scalar_1])
print(struct_3.getString())

#create instruct
instruct_2 = optrace.instruct("A123", 1, 3, "aten::sample_add", "torch", "2_11", [tensor_1, scalar_1, scalar_2, struct_3], [struct_2])
print(instruct_2.getString())
